darwin/CpuFreq.c

```diff
@@ -0,0 +1,176 @@
+    if (channels == NULL) {
+        return -1;
+    }
+
+    data->subscribed_channels = NULL;
+    data->subscription = IOReportCreateSubscription(NULL, channels, &data->subscribed_channels, 0U, NULL);
+
+    CFRelease(channels);
+
+    if (data->subscription == NULL) {
+        return -1;
+    }
+
+    data->prev_samples = NULL;
+
+    return 0;
+}
+
+void CpuFreq_update(CpuFreqData* data) {
+    CFDictionaryRef samples = IOReportCreateSamples(data->subscription, data->subscribed_channels, NULL);
+    if (samples == NULL) {
+        return;
+    }
+
+    if (data->prev_samples == NULL) {
+        data->prev_samples = samples;
+        return;
+    }
+
+    /* Residency is cumulative, we have to diff two samples to get a current view */
+    CFDictionaryRef samples_delta = IOReportCreateSamplesDelta(data->prev_samples, samples, NULL);
+
+    /* Iterate through performance state residencies. Index 0 is the idle residency, index 1-n is the per-performance state residency. */
+    __block uint32_t cpu_i = 0U;
+    IOReportIterate(samples_delta, ^(IOReportSampleRef ch) {
+        if (cpu_i >= data->existingCPUs) {
+            /* The report contains more CPUs than we know about. This should not happen. */
+            return kIOReportIterOk; // TODO: find way to possibly stop iteration early on error
+        }
+        const CpuFreqPowerStateFrequencies* cpu_frequencies = &data->cpu_frequencies_per_cluster_type[data->cluster_type_per_cpu[cpu_i]];
+        uint32_t state_count = IOReportStateGetCount(ch);
+        if (state_count != cpu_frequencies->num_frequencies + 1) {
+            /* The number of reported states does not match the number of previously queried frequencies. This should not happen. */
+            return kIOReportIterOk; // TODO: find way to possibly stop iteration early on error
+        }
+        /* Calculate average frequency based on residency and per-performance state frequency */
+        double average_freq = 0.0;
+        int64_t total_residency = 0U;
+        for (uint32_t i = 0U; i < state_count; i++) {
+            const int64_t residency = IOReportStateGetResidency(ch, i);
+            total_residency += residency;
+            /* We count idle as the smallest frequency */
+            average_freq += residency * cpu_frequencies->frequencies[(i == 0) ? 0 : (i - 1)];
+        }
+        average_freq /= total_residency;
+        data->frequencies[cpu_i] = average_freq / 1000000; // Convert to MHz
+
+        cpu_i++;
+        return kIOReportIterOk;
+    });
+    CFRelease(samples_delta);
+
+    CFRelease(data->prev_samples);
+    data->prev_samples = samples;
+}
+
+void CpuFreq_cleanup(CpuFreqData* data) {
+    if (data->subscription != NULL) {
+        CFRelease(data->subscription);
+    }
+    if (data->subscribed_channels != NULL) {
+        CFRelease(data->subscribed_channels);
+    }
+    if (data->prev_samples != NULL) {
+        CFRelease(data->prev_samples);
+    }
```

Comment:
```suggestion
    if (data->subscription)
        CFRelease(data->subscription);
    if (data->subscribed_channels)
        CFRelease(data->subscribed_channels);
    if (data->prev_samples)
        CFRelease(data->prev_samples);
```

Normally you would even drop the NULL check, but given that `CFRelease` is [kinda picky in that regard](https://stackoverflow.com/a/1608425), these checks need to stay … :( But we can at least reduce the noise …

Reply:
That's why they are there :) If NULL equivalency check is not the preferred style, I will refactor those along with the rest.